interceptors.go

```diff
@@ -0,0 +1,23 @@
+package sarama
+
+// ProducerInterceptor allows you to intercept (and possibly mutate) the records
+// received by the producer before they are published to the Kafka cluster.
+// https://cwiki.apache.org/confluence/display/KAFKA/KIP-42%3A+Add+Producer+and+Consumer+Interceptors#KIP42:AddProducerandConsumerInterceptors-Motivation
+type ProducerInterceptor interface {
+
+	// onSend is called when the producer message is intercepted.
+	// Please avoid modifying the message until it's safe to do so, as you'll be
+	// modifying the message reference.
+	onSend(*ProducerMessage)
+}
+
+// ConsumerInterceptor allows you to intercept (and possibly mutate) the records
+// received by the consumer before they are sent to the messages channel.
+// https://cwiki.apache.org/confluence/display/KAFKA/KIP-42%3A+Add+Producer+and+Consumer+Interceptors#KIP42:AddProducerandConsumerInterceptors-Motivation
+type ConsumerInterceptor interface {
+
+	// onConsume is called when the consumed message is intercepted.
+	// Please avoid modifying the message until it's safe to do so, as you'll be
+	// modifying the message reference.
+	onConsume(*ConsumerMessage)
+}
```

Comment:
some comments here:
I decided to use the object reference here to avoid the need to clone hence consume more memory when dealing with each msg.
The effect of this, is that an interceptor that changes the value and for some reason fails, the value will be changed on the reference, and the next interceptor will have the value changed.

I'm not sure we'll need to strictly follow java/ruby pattern here, as with c/go you know what are the effects of changing a reference.

happy to hear your thoughts

Reply:
I think that a bad interceptor shouldn't break the chain and the next interceptor should get the same message that the bad interceptor got

Reply:
Ah, is that what you meant by "when it's safe" 😄?

Right, so mutating the message is more memory efficient, and more importantly Go doesn't have any kind of "duplicate" method built-in, because it has basically nothing built in.
So if we wanted to support this we'd need to write a duplicate message for `ConsumerMessage` and `ProducerMessage` and keep them up to date as those structures change, or else we'll be silently corrupting things.

Hmm. 🤔

Reply:
And in Ruby, any function call could throw an error and the whole stack unwinds.
That's dangerous.

In go, you need to explicitly write:
```go
if value, err := someMethod; err != nil {
  return
}
```

Where you explicitly get an error, ignore it, and return early.
I assume the author of that code would realize the things they changed before the `return` are still changed, since the code doesn't allow you to return an error, and doesn't allow you to return a new value.

Hmm. 🤔

Reply:
yeah to all of those @psycotica0-shopify

I spent a few brain cycles on the signature of these methods.

like, why I would make the user return an error if basically I'd just ignore it and log it with the very basic logging system that Sarama has (one of the big Sarama issues, everything is logged on the Info log).

Reply:
I'm relying to much about how Go works, that might be a little different from the KIP specification, which might not be necessarily a bad thing